pkg/scalers/temporal_scaler.go

```diff
@@ -0,0 +1,172 @@
+	workflowservice "go.temporal.io/api/workflowservice/v1"
+	sdk "go.temporal.io/sdk/client"
+	"google.golang.org/grpc"
+	v2 "k8s.io/api/autoscaling/v2"
+	"k8s.io/metrics/pkg/apis/external_metrics"
+)
+
+const (
+	defaultTargetWorkflowLength           = 5
+	defaultActivationTargetWorkflowLength = 0
+)
+
+type temporalWorkflowScaler struct {
+	metricType v2.MetricTargetType
+	metadata   *temporalWorkflowMetadata
+	tcl        sdk.Client
+}
+
+type temporalWorkflowMetadata struct {
+	activationTargetWorkflowLength int64
+	endpoint                       string
+	namespace                      string
+	workflowName                   string
+	scalerIndex                    int
+	targetQueueSize                int64
+	metricName                     string
+}
+
+// NewTemporalWorkflowScaler creates a new instance of temporalWorkflowScaler.
+func NewTemporalWorkflowScaler(config *ScalerConfig) (Scaler, error) {
+	metricType, err := GetMetricTargetType(config)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get scaler metric type: %w", err)
+	}
+
+	meta, err := parseTemporalMetadata(config)
+	if err != nil {
+		return nil, fmt.Errorf("failed to parse Temporal metadata: %w", err)
+	}
+
+	c, err := sdk.Dial(sdk.Options{
+		HostPort: meta.endpoint,
+		ConnectionOptions: sdk.ConnectionOptions{
+			DialOptions: []grpc.DialOption{
+				grpc.WithTimeout(time.Duration(10) * time.Second),
+			},
+		},
+	})
+
+	if err != nil {
+		return nil, fmt.Errorf("failed to create Temporal client: %w", err)
+	}
+
+	return &temporalWorkflowScaler{
+		metricType: metricType,
+		metadata:   meta,
+		tcl:        c,
+	}, nil
+}
+
+// Close closes the Temporal client connection.
+func (s *temporalWorkflowScaler) Close(context.Context) error {
+	if s.tcl != nil {
+		s.tcl.Close()
+	}
+	return nil
+}
+
+// GetMetricSpecForScaling returns the metric specification for scaling.
+func (s *temporalWorkflowScaler) GetMetricSpecForScaling(context.Context) []v2.MetricSpec {
+	externalMetric := &v2.ExternalMetricSource{
+		Metric: v2.MetricIdentifier{
+			Name: s.metadata.metricName,
+		},
+		Target: GetMetricTarget(s.metricType, s.metadata.targetQueueSize),
+	}
+	metricSpec := v2.MetricSpec{
+		External: externalMetric,
+		Type:     externalMetricType,
+	}
+	return []v2.MetricSpec{metricSpec}
+}
+
+// GetMetricsAndActivity returns metrics and activity for the scaler.
+func (s *temporalWorkflowScaler) GetMetricsAndActivity(ctx context.Context, metricName string) ([]external_metrics.ExternalMetricValue, bool, error) {
+	queueSize, err := s.getQueueSize(ctx)
+	if err != nil {
+		return nil, false, fmt.Errorf("failed to get Temporal queue size: %w", err)
+	}
+
+	metric := GenerateMetricInMili(metricName, float64(queueSize))
+
+	return []external_metrics.ExternalMetricValue{metric}, queueSize > s.metadata.activationTargetWorkflowLength, nil
+}
+
+// getQueueSize returns the queue size of open workflows.
+func (s *temporalWorkflowScaler) getQueueSize(ctx context.Context) (int64, error) {
```

Comment:
This does not get an accurate queue size. You can use `.CountWorkflow` but that's only for workflows, it doesn't help with activities (and often it's activities that are the reason for needing to scale).

The proper way to scale Temporal workers is to use the `temporal_worker_task_slots_available` metric on the workers. See https://docs.temporal.io/dev-guide/worker-performance.

Reply:
It would make more sense if we consider multiple activities within a single workflow and deploy workers for each activity. However, the current scaling mechanism relies on pending workflows rather than individual activities. I plan to review the SDK documentation to explore the possibility of integrating activities into the scaling process.

Notably, "temporal_worker_task_slots_available" serves as a Prometheus metric, which could potentially be employed alongside the Prometheus scaler for those interested in scaling based on this particular metric.

Reply:
> It would make more sense if we consider multiple activities within a single workflow and deploy workers for each activity. 

I don't think it's a reasonable scaler if you don't consider activities. And I don't think the scaler is working that well if it's only for a single workflow type.

> However, the current scaling mechanism relies on pending workflows rather than individual activities.

Pending activities matter too (maybe more). Even if you were only doing pending workflows, list open workflows is paginated, you are not getting full counts. Regardless, scaling a worker based on a single workflow is not the best way to write a scaler.

> Notably, "temporal_worker_task_slots_available" serves as a Prometheus metric, which could potentially be employed alongside the Prometheus scaler for those interested in scaling based on this particular metric.

This is the metric that should be scaled on. The current scaler which doesn't include activities, only works for a single workflow type, etc is not sufficient IMO.

Reply:
I have transitioned this process to a paginated approach. Unfortunately, I haven't discovered a method to integrate activity counts into the current setup. It seems that further research is necessary to explore potential solutions in this regard.

Reply:
> I have transitioned this process to a paginated approach.

You should just use `CountWorkflow`, not list every workflow. But regardless, we do not have a way for you to easily get all pending activities from the server for a task queue across . The scaler needs to use the slots metric per the worker performance doc. Using list/count is not the best way to write the scaler.

Reply:
I have incorporated an option to assess the activity state and adjusted the scalar accordingly. While I acknowledge your suggestion to utilize slots metrics for scaling up or down.

Reply:
TBH, I don't have any knowledge about temporal, so I can't give you any extra insight about the implementation. 
@cretz , do you agree with the current implementation? As you (both) are the experts on temporal, I hope to get your consensus on the implementation

Reply:
This PR's implementation and your implementation are really different: https://github.com/kedacore/keda/pull/4721/files#diff-f59fd700aa9c39c0f77d364730bcf70d05712e8841b100cc6b2d502f5224724bR183-R190

Reply:
> @cretz , do you agree with the current implementation? As you (both) are the experts on temporal, I hope to get your consensus on the implementation

I do not. There are many problems with it. It doesn't work with activities and counting workflows does not mean scaling needs to occur nor does it account for scale down uses. Aggregates of `temporal_worker_task_slots_available` metric should be used to make scale up/down decisions at a container level (scale up when low, scale down when high.

Unfortunately we at Temporal have not had the time yet to invest in creating an official implementation here.

Reply:
@cretz Can you please review the recent changes?

Reply:
The idea of a target queue size and listing workflows is not the recommended approach to determining whether to scale up or down. We recommend using the `temporal_worker_task_slots_available` metric (with check whether worker type is activity or workflow).

Reply:
@cretz is one of the main contributors of [temporal.io SDK repo](https://github.com/temporalio/sdk-go).
I think that we should follow his recommendations at this point @Prajithp . Could you implement it? 🙏

Reply:
@JorTurFer I am unsure about the feasibility of achieving this through the CountWorkflow method, as it might not provide visibility into whether the activity is presently running or not. As recommended by him, individuals aiming to scale based on Prometheus data can make use of the query specified above.

Can we please leave this pull request open for a while? This would allow for the possibility of additional suggestions from others. In the meantime,  we will maintain our own fork and deploy it in our production environment.

Reply:
> Can we please leave this pull request open for a while?

Yeas sure, no problem at all

>  we will maintain our own fork and deploy it in our production environment

I'd suggest using external scaler/metrics api scaler instead of maintaining your own fork. I mean, KEDA can be extended using those scalers from a 3rd party service that you can develop with the code that you prefer. Using this approach instead of maintaining your own fork can brings you the option to upgrade KEDA without the hard effort of rebase it and adapt the code (as you have develop just a scaler this shouldn't be a drama, but extending is always better than modifying)